client/lib/domains/index.js

```diff
@@ -38,17 +38,29 @@ function canRegister( domainName, onComplete ) {
 	}
 
 	wpcom.undocumented().isDomainAvailable( domainName, function( serverError, data ) {
-		var errorCode;
 		if ( serverError ) {
-			errorCode = serverError.error;
-		} else if ( ! data.is_available && data.is_mappable ) {
-			errorCode = 'not_available_but_mappable';
-		} else if ( ! data.is_mappable && data.unmappability_reason ) {
-			errorCode = `mappable_but_${data.unmappability_reason}`;
-		} else if ( ! data.is_registrable ) {
-			errorCode = 'not_registrable';
-		} else if ( ! data.is_available ) {
-			errorCode = 'not_available';
+			onComplete( new ValidationError( serverError.error ) );
+			return;
+		}
+
+		const {
+			is_available: isAvailable,
+			is_mappable: isMappable,
+			is_registrable: isRegistrable,
+			unmappability_reason: unmappabilityReason
+		} = data;
+
+		let errorCode;
+		if ( ! isAvailable ) {
+			if ( isMappable ) {
+				errorCode = 'not_available_but_mappable';
+			} else if ( unmappabilityReason ) {
+				errorCode = `not_mappable_${ unmappabilityReason }`;
+			} else {
+				errorCode = 'not_mappable';
+			}
+		} else if ( ! isRegistrable ) {
+			errorCode = 'available_but_not_registrable';
```

Comment:
I agree with your comments - the flags should be passed to component as is (maybe `camelCase` them only) and let it figure out the rest. But that's a bigger change than I can take on now, especially since the same should be done for the mapping step. So I did the best I could to make the current approach readable.